db/db_impl/db_impl_compaction_flush.cc

```diff
@@ -101,12 +101,17 @@ bool DBImpl::ShouldRescheduleFlushRequestToRetainUDT(
   // alleviated if we continue with the flush instead of postponing it.
   const auto& mutable_cf_options = *cfd->GetLatestMutableCFOptions();
 
-  // Taking the status of the active Memtable into consideration so that we are
-  // not just checking if DB is currently already in write stall mode.
-  int mem_to_flush = cfd->mem()->ApproximateMemoryUsageFast() >=
-                             cfd->mem()->write_buffer_size() / 2
-                         ? 1
-                         : 0;
+  // Use the same criteria as WaitUntilFlushWouldNotStallWrites does w.r.t
+  // defining what a write stall is about to happen means. If this uses a
+  // stricter criteria, for example, a write stall is about to happen if the
+  // last memtable is 10% full, there is a possibility that manual flush could
+  // be waiting in `WaitUntilFlushWouldNotStallWrites` with the incorrect
+  // expectation that others will clear up the excessive memtables and
+  // eventually let it proceed. The others in this case won't start clearing
+  // until the last memtable is 10% full. To avoid that scenario, the criteria
+  // this uses should be the same or less strict than
+  // `WaitUntilFlushWouldNotStallWrites` does.
+  int mem_to_flush = cfd->mem()->IsEmpty() ? 0 : 1;
```

Comment:
For users of `max_write_buffer_number=2` will it almost never reschedule since the next memtable will usually be nonempty?

Reply:
Yes, that's a great point! Thank you for calling it out, I added some documentation in the `persist_user_defined_timestamps` option to highlight this. It's indeed a limitation with this approach that didn't precisely do a more permissive check during an ongoing manual flush. Since MyRocks has a setting that could get past this, I think we can live with this limitation for now, and it's possible to revisit and improve this in the future too.

Reply:
Works for me, thanks.